### PR TITLE
modernize: createRef API, functional components with hooks

### PR DESCRIPTION
**Pull request by astridx**
Pull Request for Issue # https://github.com/mapbox/mapbox-react-examples/issues/3.

### Summary of Changes

I implemented and added all 4 examples with hooks. ~I named the new examples `v2_nameOfExample` and not `nameOfExample-with-hooks` as suggested, because I find it clearer that way.~

I would be happy if you like this.

### CODE REVIEW

**Comment by tristen** [comment]
👋 @astridx this is awesome! What do you think about updating the existing examples instead of creating `v2_*` versions?

**Comment by astridx** [comment]
@tristen Thanks for your quick feedback

I think the original versions should be overwritten. I had scruples about just doing it like that. But: Someone who reads here is building something new and needs current information. 

That's why I deleted original directories in this PR and renamed my new directories.
Is that what you mean?

If these PR were merged, it would be good if the documentation https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/ was updated.

I have included a suggestion in the Basic Readme.:
https://github.com/mapbox/mapbox-react-examples/blob/c31fe8119e35293a902a83e0ecdd524b80042404/basic/README.md

**Comment by tristen** [comment]
> If these PR were merged, it would be good if the documentation [docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/) was updated.
> 
> I have included a suggestion in the Basic Readme.:
> [`c31fe81`/basic/README.md](https://github.com/mapbox/mapbox-react-examples/blob/c31fe8119e35293a902a83e0ecdd524b80042404/basic/README.md)

👍 I can make sure that's updated! 

@astridx your PR adds a lot of boilerplate code from create-react-app. I'd like to keep the code examples minimal so the code is well understood. Thoughts on removing additions like yarn, the testing scaffold & service workers?

**Comment by astridx** [comment]
@tristen  You're right. The unnecessary files and texts confuse others. I've cleaned up something. Can you find anything else I could take out?

File: basic/public/index.html [hunk 1]
```diff
@@ -1,14 +1,44 @@
...
+    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+    <!--
+      Notice the use of %PUBLIC_URL% in the tags above.
+      It will be replaced with the URL of the `public` folder during the build.
+      Only files inside the `public` folder can be referenced from the HTML.
+
+      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
```
Comment (tristen) [inline basic/public/index.html]: Can we drop these lines from each index?

File: basic/README.md [hunk 2]
```diff
@@ -18,3 +18,251 @@ function.
...
+Runs the app from http://localhost:3000
+
+
+
+
+### File Structure
+
+#### The HTML page
+
+Open the `index.html` file in the folder `basic-with-hooks`. This file was created by [Create React App](https://github.com/facebook/create-react-app). The line has been added for our example:
```
Comment (tristen) [inline basic/README.md]: Looks like references to `basic-with-hooks` should change to `basic`?

File: basic/README.md [hunk 3]
```diff
@@ -18,3 +18,251 @@ function.
...
 
 Runs the app from http://localhost:3000
+
+For more information see [Use Mapbox GL JS in a React app](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/) 
+
+
+Documentation
+---
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
```
Comment (tristen) [inline basic/README.md]: If we're stripping out most of the bootstrapping from create react app, Want to drop this line?

File: basic/README.md [hunk 4]
```diff
@@ -1,9 +1,9 @@
-Basic example
+Basic example using React Hooks
```
Comment (tristen) [inline basic/README.md]: Consider removing "using React Hooks" from all the readme's? If these examples are just being replaced with versions using hooks I don't think the extra messaging for that is necessary

File: basic/README.md [hunk 5]
```diff
@@ -18,3 +18,251 @@ function.
...
+...
+```
+
+This code creates the structure of the HTML page your users will see.
+
+#### Create the React app
+
+##### index.js
+
+Open the `index.js` file in the folder `basic-with-hooks/src`. This file was created by [Create React App](https://github.com/facebook/create-react-app).
```
Comment (tristen) [inline basic/README.md]: Remove references in this doc to Create React App?

File: basic/README.md [hunk 6]
```diff
@@ -18,3 +18,251 @@ function.
...
+
+index.js
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```js
```

To add code formatting

File: basic/README.md [hunk 7]
```diff
@@ -18,3 +18,251 @@ function.
...
+serviceWorker.unregister();
+```
+
+##### App.js
+
+Open the `App.js` file in the folder `basic-with-hooks/src`. This file was created by [Create React App](https://github.com/facebook/create-react-app), but we changed it.
+
+__app.js__
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```js
```

File: basic/README.md [hunk 8]
```diff
@@ -18,3 +18,251 @@ function.
...
+export default App;
+```
+
+##### Map.js
+
+Open the `Map.js` file in the folder `basic-with-hooks/src`. We created this file for this tutorial. So we have a clean project structure. The card is a separate component.
+
+__Map.js__
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```js
```

File: basic/README.md [hunk 9]
```diff
@@ -18,3 +18,251 @@ function.
...
+  );
+};
+export default Map;
+```
+
+__Step by Step__
+
+1. We are the first to import everything that is necessary.
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```js
```

File: basic/README.md [hunk 10]
```diff
@@ -18,3 +18,251 @@ function.
...
+...
+mapboxgl.accessToken = "MAPBOX_ACCESS_TOKEN";
+...
+```
+
+3. Then we create the function component `Map` and export it.
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```js
```

File: basic/README.md [hunk 11]
```diff
@@ -18,3 +18,251 @@ function.
...
+...
+```
+
+##### Map.css
+
+The map needs a few styling rules to render correctly. Open the `Map.css` file in the folder `basic-with-hooks/src`. We created this file for this tutorial:
+
+__Map.css__
+
+```
```
Comment (tristen) [inline basic/README.md]:
```suggestion
```css
```

**Review by tristen — CHANGES_REQUESTED** [review]
Thanks for removing much of the scaffolding from create-react-app! Left some quick thoughts for a couple of other things to change/remove

File: basic/README.md [hunk 12]
```diff
@@ -18,3 +18,251 @@ function.
     open http://localhost:3000/
 
 Runs the app from http://localhost:3000
+
+For more information see [Use Mapbox GL JS in a React app](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/) 
+
+
+Documentation
```
Comment (tristen) [inline basic/README.md]: Can this documentation actually be moved into a ticket? I can followup with the documentation team at Mapbox to make these changes!

Comment (astridx) [reply basic/README.md]: What do you mean with "Can this documentation actually be moved into a ticket?". I think this page have to be updated if we change the examples: https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/

**Comment by astridx** [comment]
I cleaned up read.me and index.php further. Is this OK?

File: basic/README.md (hunk 12 above)
Comment (tristen) [reply basic/README.md]: @astridx I agree! I just think the text for this should not be written in `basic/README.md`. I'm suggesting we pull this out into a [new ticket](https://github.com/mapbox/mapbox-react-examples/issues/new) with a title that says "Update the documentation with this text to account for new changes"?

**Comment by astridx** [comment]
@tristen Please see: https://github.com/mapbox/mapbox-react-examples/issues/12

File: basic/src/Map.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      container: mapContainerRef.current,
+      style: "mapbox://styles/mapbox/streets-v11",
+      center: [lng, lat],
+      zoom: zoom,
+    });
+
+    // Add navigation control (the +/- zoom buttons)
+    map.addControl(new mapboxgl.NavigationControl(), "top-right");
+
+    map.on("moveend", () => {
```
Comment (tristen) [inline basic/src/Map.js]: Can we keep using the `"move"` event here?

File: react-tooltip/src/components/Tooltip.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+
+const Tooltip = ({ feature }) => {
+  const { id } = feature.properties;
+  console.log(id);
+
+  return (
+    <div id={`tooltip-${id}`}>
+      <h3>Informations</h3>
```
Comment (tristen) [inline react-tooltip/src/components/Tooltip.js]: Can we drop this heading?

File: react-tooltip/src/components/Tooltip.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+      <strong>Layer:</strong> {feature.layer["source-layer"]}
```
Comment (tristen) [inline react-tooltip/src/components/Tooltip.js]:
```suggestion
      <strong>Source layer:</strong> {feature.layer["source-layer"]}
```

File: react-tooltip/src/components/Tooltip.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+      <br />
+      <strong>ID:</strong> {feature.layer.id}
```
Comment (tristen) [inline react-tooltip/src/components/Tooltip.js]:
```suggestion
      <strong>Layer ID:</strong> {feature.layer.id}
```

File: react-tooltip/src/components/Tooltip.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+
+const Tooltip = ({ feature }) => {
+  const { id } = feature.properties;
+  console.log(id);
```
Comment (tristen) [inline react-tooltip/src/components/Tooltip.js]: Can you drop this?

File: react-tooltip/public/index.html [hunk 18]
```diff
@@ -1,14 +1,13 @@
 <!DOCTYPE html>
-<html lang='en'>
-<head>
-  <title></title>
-  <meta charset='utf-8'>
-  <meta name='viewport' content='width=device-width, initial-scale=1'>
-  <link href='https://api.mapbox.com/mapbox-assembly/mbx/v0.18.0/assembly.min.css' rel='stylesheet'>
-  <link href='https://api.tiles.mapbox.com/mapbox-gl-js/v0.39.1/mapbox-gl.css' rel='stylesheet' />
```
Comment (tristen) [inline react-tooltip/public/index.html:8]: For all the examples, can we keep the references to these scripts?

File: react-tooltip/public/index.html [hunk 19]
```diff
@@ -1,14 +1,13 @@
...
-</head>
-<body>
-  <div id='app'></div>
-  <script src='https://api.mapbox.com/mapbox-assembly/mbx/v0.18.0/assembly.js'></script>
```
Comment (tristen) [inline react-tooltip/public/index.html:12]: For all the examples, can we keep the reference to this script?

File: react-tooltip/README.md [hunk 20]
```diff
@@ -1,5 +1,4 @@
 React powered tooltip
----
```
Comment (tristen) [inline react-tooltip/README.md]: Can we keep this? It's markdown to turn "React powered tooltip" into a heading

File: data-overlay/README.md [hunk 21]
```diff
@@ -1,5 +1,4 @@
 Data overlay example
----
```
Comment (tristen) [inline data-overlay/README.md]: Can we keep this? It's markdown to turn "Data overlay example" into a heading

File: data-overlay-redux/src/Map.js [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+        "\n",
+        {},
+        ["get", "name"],
+        {
+          "font-scale": 0.8,
+          "text-font": [
+            "literal",
+            ["DIN Offc Pro Italic", "Arial Unicode MS Regular"],
+          ],
+        },
```
Comment (tristen) [inline data-overlay-redux/src/Map.js]: Nice addition!

**Review by tristen — CHANGES_REQUESTED** [review]
Thanks so much for the changes so far and your patience. I left a more through code review, then this should be good to go after those changes.

**One broad change to make.** Can you change [all strings to use single quotes](https://github.com/airbnb/javascript#strings--quotes)?

---

I'm also not seeing the controls to switch between data and the legend in the `data-overlay` and `data-overlay-react` examples:

Screen Shot 2020-08-27 at 10 30 20 AM

**Comment by astridx** [comment]
@tristen I am glad that you look at it and improve it.
I hope I've found everything you suggested. Sorry about the not working examples. The controls were not displayed because I accidentally deleted the CSS file. But now I've tested everything again. All examples should work.

**Review by tristen — APPROVED** [review]
@astridx the latest changes look great! I'm going to go ahead and merge this and follow up on documentation in #12

**Pull request merged**